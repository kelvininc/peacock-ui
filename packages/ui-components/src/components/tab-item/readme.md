# *<kv-tab-item>*



<!-- Auto Generated Below -->


## Usage

### React

```tsx
import React from 'react';
import { KvTabItem } from '@kelvininc/react-ui-components';

export const TabItemExample: React.FC = () => (
	<>
		{/*-- With required props --*/}
		<KvTabItem tabKey="tab1" label="Tab 1" />

		{/*-- Disabled --*/}
		<KvTabItem tabKey="tab1" label="Tab 1" disabled></KvTabItem>

		{/*-- Selected --*/}
		<KvTabItem tabKey="tab1" label="Tab 1" selected></KvTabItem>

		{/*-- With Notification --*/}
		<KvTabItem tabKey="tab1" label="Tab 1" hasNotification></KvTabItem>

		{/*-- With Notification Color --*/}
		<KvTabItem tabKey="tab1" label="Tab 1" hasNotification notificationColor="#202020"></KvTabItem>
	</>
);
```



## Properties

| Property              | Attribute            | Description                                                                                                                                                                             | Type                                           | Default                |
| --------------------- | -------------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ---------------------------------------------- | ---------------------- |
| `customClass`         | `custom-class`       | (optional) Additional classes to apply for custom CSS. If multiple classes are provided they should be separated by spaces. It is also valid to provide CssClassMap with boolean logic. | `CssClassMap \| string \| string[]`            | `''`                   |
| `customStyle`         | --                   | (optional) Additional style to apply for custom CSS.                                                                                                                                    | `{ [key: string]: string; }`                   | `undefined`            |
| `disabled`            | `disabled`           | (optional) To disable this tab                                                                                                                                                          | `boolean`                                      | `false`                |
| `hasNotification`     | `has-notification`   | (optional) To show/hide notification icon or not                                                                                                                                        | `boolean`                                      | `false`                |
| `icon`                | `icon`               | (optional) The tab's icon                                                                                                                                                               | `EIconName \| EOtherIconName`                  | `undefined`            |
| `label` _(required)_  | `label`              | (required) Name to show in UI for this tab                                                                                                                                              | `string`                                       | `undefined`            |
| `notificationColor`   | `notification-color` | (optional) The tab's notification color (hex value, rgb or css var format)                                                                                                              | `string`                                       | `''`                   |
| `selected`            | `selected`           | (optional) To set this tab as the selected one                                                                                                                                          | `boolean`                                      | `false`                |
| `size`                | `size`               | (optional) Sets this tab item to a different styling configuration                                                                                                                      | `EComponentSize.Large \| EComponentSize.Small` | `EComponentSize.Large` |
| `tabKey` _(required)_ | `tab-key`            | (required) A unique identifier for this tab                                                                                                                                             | `number \| string`                             | `undefined`            |


## Events

| Event         | Description                      | Type                            |
| ------------- | -------------------------------- | ------------------------------- |
| `tabSelected` | Emitted when the tab is selected | `CustomEvent<number \| string>` |


## Shadow Parts

| Part     | Description          |
| -------- | -------------------- |
| `"icon"` | The tab's item icon. |


## CSS Custom Properties

| Name                    | Description                       |
| ----------------------- | --------------------------------- |
| `--tab-item-icon-color` | The color of tab navigation icon; |
| `--tab-item-icon-size`  | The size of tab navigation icon;  |


## Dependencies

### Used by

 - [kv-tab-navigation](../tab-navigation)

### Depends on

- [kv-icon](../icon)

### Graph
```mermaid
graph TD;
  kv-tab-item --> kv-icon
  kv-tab-navigation --> kv-tab-item
  style kv-tab-item fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------


