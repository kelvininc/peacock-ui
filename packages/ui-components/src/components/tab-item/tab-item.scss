@import '../../assets/styles/globals';

$tab-dot-error-size: 6px;

:host {
	height: inherit;

	.tab-item-container {
		height: 100%;
		display: flex;
		align-items: center;
		margin: 0 $spacing-6x;
		user-select: none;
		white-space: nowrap;
		cursor: pointer;

		&.selected .label {
			color: kv-color('text');
		}

		&.disabled {
			cursor: not-allowed;
			pointer-events: none;

			.label {
				color: kv-color('text', 'step-5');
			}
		}

		.label {
			@include kv-font-h4-semibold();
			color: kv-color('text', 'step-4');
		}

		&.small {
			margin: 0 $spacing-4x;

			.label {
				@include kv-font-label-small-semibold();
			}
		}

		.notification-dot {
			width: $tab-dot-error-size;
			height: $tab-dot-error-size;
			border-radius: 50%;
			margin-left: $spacing-2x;
			background-color: kv-color('error');
			animation: circleGrow 0.25s;
		}
	}

	@keyframes circleGrow {
		from {
			transform: scale(0);
		}

		to {
			transform: scale(1);
		}
	}
}
