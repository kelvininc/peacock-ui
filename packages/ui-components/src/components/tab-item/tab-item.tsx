import { Component, Host, h, Prop, EventEmitter, Event } from '@stencil/core';
import { throttle } from 'lodash-es';
import { EComponentSize } from '../../utils/types';

@Component({
	tag: 'kv-tab-item',
	styleUrl: 'tab-item.scss',
	shadow: true
})
export class KvTabItem {
	/** (required) A unique identifier for this tab */
	@Prop() tabKey!: number | string;
	/** (required) Name to show in UI for this tab */
	@Prop() label!: string;
	/** (optional) To disable this tab */
	@Prop() disabled?: boolean = false;
	/** (optional) To set this tab as the selected one */
	@Prop() selected?: boolean = false;
	/** (optional) To show/hide notification icon or not */
	@Prop() hasNotification: boolean = false;
	/** (optional) The tab's notification color (hex value, rgb or css var format) */
	@Prop() notificationColor?: string = '';
	/** (optional) Sets this tab item to a different styling configuration */
	@Prop() size?: EComponentSize = EComponentSize.Large;
	/** Emitted when the tab is selected */
	@Event() tabSelected: EventEmitter<number | string>;

	private tabClickThrottler: () => void;

	connectedCallback() {
		this.tabClickThrottler = throttle(() => this.tabClick(), 300);
	}

	private tabClick = () => {
		this.tabSelected.emit(this.tabKey);
	};

	render() {
		const customStyles = {
			backgroundColor: this.notificationColor
		};

		return (
			<Host>
				<div
					class={{
						'tab-item-container': true,
						'selected': this.selected,
						'disabled': this.disabled,
						'has-notification': this.hasNotification,
						'small': this.size === EComponentSize.Small
					}}
					onClick={this.tabClickThrottler}
				>
					<div class="label">{this.label}</div>
					{this.hasNotification && <div class="notification-dot" style={customStyles}></div>}
				</div>
			</Host>
		);
	}
}
