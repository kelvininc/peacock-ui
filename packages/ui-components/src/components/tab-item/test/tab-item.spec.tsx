import { newSpecPage, SpecPage } from '@stencil/core/testing';
import { KvTabItem } from '../tab-item';

describe('Tab Item (unit tests)', () => {
	let page: SpecPage;
	let component: KvTabItem;

	describe('when initialized with required props', () => {
		beforeEach(async () => {
			page = await newSpecPage({
				components: [KvTabItem],
				html: `<kv-tab-item tab-key="dashboard" label="Dashboard"></kv-tab-item>`
			});
			component = page.rootInstance;
		});

		it('should match the snapshot', () => {
			expect(page.root).toMatchSnapshot();
		});

		it('should set the tabKey prop value', () => {
			expect(component.tabKey).toEqual('dashboard');
		});

		it('should set the label prop value', () => {
			expect(component.label).toEqual('Dashboard');
		});
	});

	describe('when initialized with disabled attribute', () => {
		beforeEach(async () => {
			page = await newSpecPage({
				components: [KvTabItem],
				html: `<kv-tab-item tab-key="dashboard" label="Dashboard" disabled></kv-tab-item>`
			});
			component = page.rootInstance;
		});

		it('should match the snapshot', () => {
			expect(page.root).toMatchSnapshot();
		});

		it('should set the disabled prop value', () => {
			expect(component.disabled).toBeTruthy();
		});
	});

	describe('when initialized with selected attribute', () => {
		beforeEach(async () => {
			page = await newSpecPage({
				components: [KvTabItem],
				html: `<kv-tab-item tab-key="dashboard" label="Dashboard" selected></kv-tab-item>`
			});
			component = page.rootInstance;
		});

		it('should match the snapshot', () => {
			expect(page.root).toMatchSnapshot();
		});

		it('should set the disabled prop value', () => {
			expect(component.selected).toBeTruthy();
		});
	});

	describe('when initialized with notification-color attribute', () => {
		beforeEach(async () => {
			page = await newSpecPage({
				components: [KvTabItem],
				html: `<kv-tab-item tab-key="dashboard" label="Dashboard" notification-color="#EFEFEF"></kv-tab-item>`
			});
			component = page.rootInstance;
		});

		it('should match the snapshot', () => {
			expect(page.root).toMatchSnapshot();
		});

		it('should set the notificationColor prop value', () => {
			expect(component.notificationColor).toEqual('#EFEFEF');
		});
	});
});
