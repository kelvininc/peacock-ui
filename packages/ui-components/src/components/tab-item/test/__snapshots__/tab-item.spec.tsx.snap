// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tab Item (unit tests) when initialized with disabled attribute should match the snapshot 1`] = `
<kv-tab-item disabled="" label="Dashboard" tab-key="dashboard">
  <mock:shadow-root>
    <div class="disabled tab-item-container">
      <div class="label">
        Dashboard
      </div>
    </div>
  </mock:shadow-root>
</kv-tab-item>
`;

exports[`Tab Item (unit tests) when initialized with notification-color attribute should match the snapshot 1`] = `
<kv-tab-item label="Dashboard" notification-color="#EFEFEF" tab-key="dashboard">
  <mock:shadow-root>
    <div class="tab-item-container">
      <div class="label">
        Dashboard
      </div>
    </div>
  </mock:shadow-root>
</kv-tab-item>
`;

exports[`Tab Item (unit tests) when initialized with required props should match the snapshot 1`] = `
<kv-tab-item label="Dashboard" tab-key="dashboard">
  <mock:shadow-root>
    <div class="tab-item-container">
      <div class="label">
        Dashboard
      </div>
    </div>
  </mock:shadow-root>
</kv-tab-item>
`;

exports[`Tab Item (unit tests) when initialized with selected attribute should match the snapshot 1`] = `
<kv-tab-item label="Dashboard" selected="" tab-key="dashboard">
  <mock:shadow-root>
    <div class="selected tab-item-container">
      <div class="label">
        Dashboard
      </div>
    </div>
  </mock:shadow-root>
</kv-tab-item>
`;
